### DISCUSSION TITLE
Is it safe to pass matrices for non-affine transformations to e.g. `glm_translate`?

### DISCUSSION THREAD

**Discussion by v1993** [recp/cglm#370] [Q&A] [answered]
I'm trying to determine if functions from `affine-pre/post.h` headers can be reasonably added directly to `mat4` class in my Vala bindings or not. So, are affine transformation functions (save for those from `cglm/affine-mat.h`) safe to use on matrices with a projection part? The answer appears to be yes, but I'd just like to be sure.

**Comment by recp** [comment, marked as answer]
Hi @v1993,

Affine functions can be used with any matrices ( if there is a need ) but `affine-mat.h` cannot, because `affine-mat.h` has assumptions like:

```C
 * Matrix format should be:
 *   R  R  R  X
 *   R  R  R  Y
 *   R  R  R  Z
 *   0  0  0  W
```

```C
 * Right Matrix format should be:
 *   R  R  R  0
 *   R  R  R  0
 *   R  R  R  0
 *   0  0  0  1
```

```
inverse orthonormal rotation + translation matrix (ridig-body)
```

Actually there are some functions in affine too that needs specific type of matrix:

```
 * @brief decompose affine transform, TODO: extract shear factors.
 *        DON'T pass projected matrix here
```

**Reply by v1993** [reply]
Great, thanks for confirming this! Another quick question: is lack of `ivec` functions from struct API intended to be fixed or are those functions considered deprecated?

**Reply by recp** [reply]
`is lack of ivec functions` for instance? I'm not sure we should bring all vec3 functions to ivec because of int <-> float, feedbacks are welcome of course

**Reply by v1993** [reply]
There are no ivec headers (and therefore functions) in struct API at all, that's what I'm talking about.

**Reply by recp** [reply]
Wow, sorry just realized that there is no ivec in struct api at all except print funcs! I think we should bring all array api to struct including ivec
